Brooks/Helpers.hs

```diff
@@ -0,0 +1,159 @@
+    hmap,himap,hicat,vicat,vibcat,vmap,vimap,vibmap, reduceLibs,
+    tripFst, tripSnd, tripThird, liftA4, liftA5, liftA6, liftA7, liftA8,
+    liftList, lift2Lists, lift1List, liftPair, lift3Pair, lift4Pair, 
+    liftPairFst, liftTripFst
+) where
+
+import Prelude hiding ((<>))
+import Control.Monad (filterM)
+import Control.Applicative (liftA2, liftA3)
+import Data.Char (toUpper)
+import Data.String.Utils (replace)
+import Data.List (nub,intersperse)
+import Text.PrettyPrint.HughesPJ (Doc, vcat, hcat, text, char, doubleQuotes, 
+  (<>), ($+$), comma, punctuate, nest)
+
+blank :: Doc
+blank = text ""
+
+spc :: Doc
+spc = text " "
+
+oneTabbed :: [Doc] -> Doc
+oneTabbed = vcat . map oneTab
+
+oneTab :: Doc -> Doc
+oneTab = nest 4
+
+vertical :: (a -> Doc) -> [a] -> Doc
+vertical f = vcat . map f
+
+verticalComma :: (a -> Doc) -> [a] -> Doc
+verticalComma f = vcat . punctuate comma . map f
+
+verticalNewLine :: (a -> Doc) -> [a] -> Doc
+verticalNewLine f = vcat . punctuate (blank $+$ blank) . map f
+
+angles :: Doc -> Doc
+angles d = char '<' <> d <> char '>'
+
+doubleQuotedText :: String -> Doc
+doubleQuotedText = doubleQuotes . text
+
+doubleQuoted :: (a -> String) -> a -> Doc
+doubleQuoted labeller = doubleQuotedText . labeller
+
+capitalize :: String -> String
+capitalize [] = error "capitalize called on an emptry String"
+capitalize (c:cs) = (toUpper c):cs
+
+containsAll :: Eq a => [a] -> [a] -> Bool
+containsAll = all . flip elem
+
+makeLiteralNameValid :: String -> String
+makeLiteralNameValid s =
+    let chars = ["\\","\""]
+    in foldl myLiteralNameReplace s chars
+
+makeVarNameValid :: String -> String
+makeVarNameValid s =
+    let illegalChars = [
+            "~","`","-","=","!","@","#","$","%","^","&","*","(",")","+",
+            "[","]","\\",";","'",".","/","{","}","|",":","\"","<",">","?"," "]
+    in foldl myVarNameReplace s illegalChars
+
+makeClassNameValid :: String -> String
+makeClassNameValid = makeVarNameValid
+
+--http://community.livejournal.com/evan_tech/220036.html
+powerSet :: Eq a => [a] -> [[a]]
+powerSet = filterM (const [True, False]) . nub
+
+hmap :: (a -> Doc) -> [a] -> Doc
+hmap f = hcat . map f
+
+himap :: Doc -> (a -> Doc) -> [a] -> Doc
+himap c f = hcat . intersperse c . map f
+
+hicat :: Doc -> [Doc] -> Doc
+hicat c l = hcat $ intersperse c l
+
+vicat :: Doc -> [Doc] -> Doc
+vicat c = vcat . intersperse c
+
+vibcat :: [Doc] -> Doc
+vibcat = vicat blank
+
+vmap :: (a -> Doc) -> [a] -> Doc
+vmap f = vcat . map f
+
+vimap :: Doc -> (a -> Doc) -> [a] -> Doc
+vimap c f = vicat c . map f
+
+vibmap :: (a -> Doc) -> [a] -> Doc
+vibmap = vimap blank
+
+reduceLibs :: [String] -> [String] -> [String]
+reduceLibs libs modules = nub $ filter (`notElem` modules) libs 
+
+tripFst :: (a, b, c) -> a
+tripFst (c, _, _) = c
+
+tripSnd :: (a, b, c) -> b
+tripSnd (_, n, _) = n
+
+tripThird :: (a, b, c) -> c
+tripThird (_, _, b) = b
+
+liftA4 :: Applicative f => (a -> b -> c -> d -> e) -> f a -> f b -> f c -> f d -> f e
```

Comment:
making sure we actually use all of these would be worthwhile too.

Reply:
Checked all of the helpers and removed unused ones in 3c6f8c9cba941ab386ef6ba8fcb6ea0d88d0cdc4.